src/core/MemoryManager.cpp

```diff
@@ -25,197 +24,54 @@
-	s_memoryPools.append( m );
-	return true;
-}
-
-
-void * MemoryManager::alloc( size_t size )
-{
-	if( !size )
-	{
-		return NULL;
+/// Global static object handling rpmalloc intializing and finalizing
+struct MemoryManagerGlobalGuard {
+	MemoryManagerGlobalGuard() {
+		rpmalloc_initialize();
 	}
-
-	int requiredChunks = size / MM_CHUNK_SIZE + ( size % MM_CHUNK_SIZE > 0 ? 1 : 0 );
-
-	MemoryPool * mp = NULL;
-	void * ptr = NULL;
-
-	MemoryPoolVector::iterator it = s_memoryPools.begin();
-
-	s_poolMutex.lockForRead();
-	while( it != s_memoryPools.end() && !ptr )
-	{
-		ptr = ( *it ).getChunks( requiredChunks );
-		if( ptr )
-		{
-			mp = &( *it );
-		}
-		++it;
-	}
-	s_poolMutex.unlock();
-
-	if( ptr )
-	{
-		s_pointerMutex.lock();
-		PtrInfo p;
-			p.chunks = requiredChunks;
-			p.memPool = mp;
-		s_pointerInfo[ptr] = p;
-		s_pointerMutex.unlock();
-		return ptr;
+	~MemoryManagerGlobalGuard() {
+		rpmalloc_finalize();
 	}
+} static mm_global_guard;
 
-	// can't find enough chunks in existing pools, so
-	// create a new pool that is guaranteed to have enough chunks
-	int moreChunks = qMax( requiredChunks, MM_INCREMENT_CHUNKS );
-	int i = MemoryManager::extend( moreChunks );
 
-	mp = &s_memoryPools[i];
-	ptr = s_memoryPools[i].getChunks( requiredChunks );
-	if( ptr )
-	{
-		s_pointerMutex.lock();
-		PtrInfo p;
-			p.chunks = requiredChunks;
-			p.memPool = mp;
-		s_pointerInfo[ptr] = p;
-		s_pointerMutex.unlock();
-		return ptr;
-	}
-	// still no luck? something is horribly wrong
-	qFatal( "MemoryManager.cpp: Couldn't allocate memory: %d chunks asked", requiredChunks );
-	return NULL;
+namespace {
+static thread_local size_t thread_guard_depth;
 }
 
-
-void MemoryManager::free( void * ptr )
+MemoryManager::ThreadGuard::ThreadGuard()
 {
-	if( !ptr )
-	{
-		return; // Null pointer deallocations are OK but do not need to be handled
+	if (thread_guard_depth++ == 0) {
+		rpmalloc_thread_initialize();
 	}
-
-	// fetch info on the ptr and remove
-	s_pointerMutex.lock();
-	if( ! s_pointerInfo.contains( ptr ) ) // if we have no info on ptr, fail loudly
-	{
-		qFatal( "MemoryManager: Couldn't find pointer info for pointer: %p", ptr );
-	}
-	PtrInfo p = s_pointerInfo[ptr];
-	s_pointerInfo.remove( ptr );
-	s_pointerMutex.unlock();
-
-	p.memPool->releaseChunks( ptr, p.chunks );
-}
-
-
-int MemoryManager::extend( int chunks )
-{
-	MemoryPool m ( chunks );
-	m.m_pool = MemoryHelper::alignedMalloc( chunks * MM_CHUNK_SIZE );
-
-	s_poolMutex.lockForWrite();
-	s_memoryPools.append( m );
-	int i = s_memoryPools.size() - 1;
-	s_poolMutex.unlock();
-
-	return i;
 }
 
-
-void MemoryManager::cleanup()
+MemoryManager::ThreadGuard::~ThreadGuard()
 {
-	for( MemoryPoolVector::iterator it = s_memoryPools.begin(); it != s_memoryPools.end(); ++it )
-	{
-		MemoryHelper::alignedFree( ( *it ).m_pool );
-		MemoryHelper::alignedFree( ( *it ).m_free );
+	if (--thread_guard_depth == 0) {
+		rpmalloc_thread_initialize();
```

Comment:
You probably meant to do `rpmalloc_thread_finalize` here

Reply:
Absolutely. Thanks!